### ISSUE TITLE
Separate Campaign database into a separate object

### ISSUE THREAD

**Issue by dww100** [UCL-CCS/EasyVVUQ#72]
Okay so my thought here is that it maybe that we want to provide the same database interface to both the `Campaign` and `Worker` parts of VVUQ workflows.

The idea being that eventually we can separate concerns such that:
- `Campaign` deals with book keeping (logging operations such as sampling etc.) and validation (as such needs to store parameter space description and sampler description, along with access to `Database`).
- `Database` stores application specific information (Encoder, Decoder, execution command), collation information (target directory for decoded data, aggregation method) and individual Run parameters.

This might in addition mandate a change in the existing database plan (i.e. needs adding a collation table).

As such interface to `Database` would need:

**Setters**
app
collation_info(campaign_dir, collation_method)
add_run

**Getters**
app
encoder : convenience - info in app table
decoder : convenience - info in app table
execution :  convenience - info in app table
get_run(run_id)
list_incomplete_runs

Another part of this redesign would be related to the `Sampler`s and selection of variables for inclusion in the sampling. We need to have two lists here variables which are 'variable' in the sense that the user potentially wants it varied and 'varying' meaning currently being assessed through sampling. The the former should be stored in `Campaign` directly and the latter in the `Sampler` (which should log its serialised form in `Campaign` upon update).

Thoughts?

**Comment by dww100** [comment]
As part of the WP2 and QCG interactions and this I have been working on defining some terms:
https://github.com/UCL-CCS/EasyVVUQ/wiki/Proposed-terminology

**Comment by dww100** [comment]
Note for myself: in the long run may have multiple apps

**Comment by dww100** [comment]
Current proposed reworking of the database structure: https://dbdiagram.io/d/5cc0422cf7c5bb70c72fbd64

**Comment by djgroen** [comment]
@dww100 The design makes sense to me in general. The one thing that I think could become hairy is the "status" field in the Run object. Different backends could create inconsistencies in what that field would contain.

Perhaps it's best to go for a very strict and simplistic convention there, and force backends to keep any other data within their own states?

**Comment by dww100** [comment]
@djgroen Is your issue just with the name? To be honest I would rather backends created their own `run_status` table, but can see a case for a change to something like `vvuq_status` for clarity. It will be an issue that we won't be able to completely police how people edit table entries external to the Campaign interface.

**Comment by djgroen** [comment]
@dww100 No, I think the name is okay. It's just that I think we should tightly define what the value of run_status could be, to avoid spagbollification. Perhaps we could even make it an enum or something?

e.g. one of {'planned','in progress','completed'} ?

**Comment by dww100** [comment]
@djgroen I thought that ENUM was not an SQL standard? Seems there is SQLAlchemy documentation though https://docs.sqlalchemy.org/en/13/core/type_basics.html

To be honest I don't want backends touching it at all, so limiting how they screw things up seems only a minor concern. The execution step is outside the concerns of EasyVVUQ. This is the EasyVVUQ status.

**Comment by djgroen** [comment]
@dww100 Ah, okay. If it's not used by external tools, then it makes sense to keep it as a string, as it's more standard :).

I guess there is a mechanism how the EasyVVUQ status will be updated, based on the execution outcome... I'm not entirely sure how that part will work at this point, but I guess that might become clear further down the line :).

**Comment by dww100** [comment]
It's based on whether the decoder has successfully run on the output.

**Comment by djgroen** [comment]
Ah, that sounds ideal then :).
